cpp/src/hdbscan/detail/utils.h

```diff
@@ -212,6 +213,42 @@ void normalize(value_t* data, value_idx n, size_t m, cudaStream_t stream)
     stream);
 }
 
+/**
+ * Computes softmax (unnormalized) of membership vector.
+ * @tparam value_idx
+ * @tparam value_t
+ * @param[in] handle raft handle for resource reuse
+ * @param[inout] condensed_tree cluster tree (condensed hierarchy with all nodes of size > 1)
+ * @param[in] sorted_parents parents array sorted
+ * @param[out] indptr CSR indptr of parents array after sort
+ */
+template <typename value_idx, typename value_t>
+void softmax(const raft::handle_t& handle, value_t* data, value_idx n, size_t m)
```

Comment:
We have to create a copy of the data in order to avoid inconsistency during normalization (for numerical stability)

Reply:
That doesn't sound right- we shouldn't have to make a copy here. If these operations are being scheduled wrt the same cuda stream, it should be guaranteed that any prior operations scheduled on compute resources are complete before new operations are scheduled.

One thing we can do to test the theory that there's a data race somewhere is to use the input directly (instead of copying) and call `handle.sync_stream()` before and after computing the argmax. Allocating more memory will also cause a device synchronization, which would have the same effect as doing a stream sync

Reply:
What I meant was that if this operation:
`data[idx] = exp(data_copy[idx] - data_copy[n*(idx / n) + membership_argmax[idx / n]])`
was replaced by this:
`data[idx] = exp(data[idx] - data[n*(idx / n) + membership_argmax[idx / n]])`, then it is possible that `data[n*(idx / n) + membership_argmax[idx / n]]` got modified while/before `data[idx]` was being computed since they are both being modified in parallel in the same `thrust::for_each` invocation.

Reply:
I don't think you need to copy the whole dataset to do this, though. It looks like what you want is to subtract the value in a specific column for each row from each value in `data`, right? Instead of copying all `m*n` entries from the input, why not just create an array of size of `m` that contains all the max values for each row? That way you've reduced your problme down to the number of rows.

Reply:
Also, I would highly suggest using `raft::matrix::map_offset` primitive in raft for this. The benefit to using the raft primitives (even if some of them end up just wrapping thrust calls) is that it provides a facade layer for future optimizations.